### ISSUE TITLE
WaitGroup can try to call the callback before you've ever called wait

### ISSUE THREAD

**Issue by azylman** [dtjm/waitgroup.js#3]
`waitgroup.js` doesn't totally mirror the Go interface for `WaitGroup`s, since the callback is invoked from the `done` method instead of the `wait` method. If all `done` methods are called before `wait` is called, an exception is thrown.

Here is the example program, modified in such a way that it's possible to throw an exception:

```javascript
var wg = new WaitGroup;

wg.add();
initResource1().on("done", function(){
  wg.done();
});

wg.add();
initResource2().on("done", function(){
  wg.done();
});

wg.add();
initResource3().on("done", function(){
  wg.done();
});

setTimeout(function(){
  wg.wait(function(){
    console.log("All done!");
  })},
100);
```

Having the callback invoked from the wait has another benefit that it would also allow you to pass your `WaitGroup` around and wait for the same tasks to complete in multiple places (something else that works in Go's `WaitGroup`, but not this one).

An easy way to do this would be to make the `done` method only increment `this.ready` and have the `wait` method wait until `this.ready === this.total`, and then invoke the function that was passed in.

**Comment by dtjm** [comment]
@azylman Thanks for the report. Would you be able to submit a pull request? I might not be able to get to this soon.

**Comment by azylman** [comment]
I'll do it if I'm able. There's a complication, though, that I'm not sure how to handle. It looks like there's an undocumented feature of passing in the callback through the constructor, which would enable things like:

```javascript
wg = new WaitGroup(function(){
  console.log("All done!");
});

wg.add()
initResource().on("done", function(){
  wg.done()
});
```

My proposed solution would break that use case - I'm not sure if you care, though, since it's undocumented? Let me know how you'd like to proceed and I'll see if I get a chance to submit a PR.

**Comment by dtjm** [comment]
That's ok if you remove that functionality, I think it's unintentionally there from when I initially created this library.

**Comment by dtjm** [comment]
merged #4
